File: programmers/DFS_BFS/타겟 넘버.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int target_num;
int leng;
int cnt;

void make_number(int idx, int buf, const vector<int> & numbers)
{
    if(idx == leng)
    {
        if(buf == target_num)
        {
            ++cnt;
        }
        return;
    }
    
    make_number(idx + 1, buf + numbers[idx], numbers);
    make_number(idx + 1, buf - numbers[idx], numbers);
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    target_num = target;
    leng = numbers.size();
    
    make_number(0, 0, numbers);
    answer = cnt;
    return answer;
}

File: programmers/DFS_BFS/타겟_넘버.java
package com.kt.aicare.common.util.programmers.DFS_BFS;

public class 타겟_넘버 {
  int goal = 0;
  int cnt = 0;
  void initData(int target) {
    goal = target;
    cnt = 0;
  }

  void generateHelper(int curIdx, int cache, int[] numbers) {
    if (curIdx == numbers.length) {
      if (cache == goal) {
        cnt++;
      }
      return;
    }

    generateHelper(curIdx + 1, cache + numbers[curIdx], numbers);
    generateHelper(curIdx + 1, cache - numbers[curIdx], numbers);
  }

  public int solution(int[] numbers, int target) {
    initData(target);
    generateHelper(0, 0, numbers);
    return cnt;
  }

}
